arm: core: cortex_m: add a barrier before the dummy FP instruction

aurel32 opened the pull request:
On Cortex-M7 CPU (at least on STM32F723 with patches from https://github.com/zephyrproject-rtos/zephyr/pull/7284), the dummy move FPU instruction is executed before the FPU lazy state preservation is disabled. This patch adds a barrier before it to avoid that.

This can be easily reproduced with a simple example, like `hello_world`, by setting `CONFIG_FLOAT=y`. The system goes in an endless loop very early, before outputting anything on the console. This is something difficult to debug, as executing the instructions step by step in GDB hides the problem. I guess the other Cortex-M7 based platforms are also affected: NXP i.MX RT and Microchip SAM E70. I guess @MaureenHelm can test the first one.

galak commented:
Do we need both isb & dsb?

ioannisg commented:
Good point, I would think that only the instruction barrier is required.

aurel32 commented:
> Do we need both isb & dsb?

I have to say I copied the barriers from below. From what I understand we need both:
- dsb to make sure the write to FPCCR is finished
- isb to empty the pipeline and execute vmov using the new FPU configuration

ioannisg commented:
IMHO, a DSB instruction would be needed if we are doing read/write to RAM or system registers and then use that value (so make sure the value is updated before performing new access).
Here, all we need to guarantee is that the FPCCR is updated before we do the  dummy write, so we only need to preserve execution synchronization (but we do not use the  value of the FPCCR anywhere).

aurel32 commented:
Ok, I think I got it. Given your explanation I guess it also means the dsb after the vmov is useless.

aurel32 committed:
arch: arm: core: cortex_m: add a barrier before the dummy FP instruction

On Cortex-M7 CPU (at least on STM32F723), the dummy move FPU instruction
is executed before the FPU lazy state preservation is disabled. Add an
instruction synchronization barrier before it to avoid that.

At the same time, remove the data synchronization barrier after the
dummy move as it does not have any effect on RAM or registers.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
I have just pushed a new version without the dsb instruction.

agross-oss reviewed (approved):
with the dsb removal this looks ok

ioannisg reviewed (approved):
Looks OK to me.

ioannisg commented:
I think the commit message prefix should read: "arch: arm: ......."

aurel32 commented:
I have just updated the patch to add the "arch:" prefix.
